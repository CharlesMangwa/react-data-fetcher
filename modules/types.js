/* @flow */
/* eslint-disable no-use-before-define */

import PropTypes from 'prop-types'

// FLOW

export type Interceptor = (InterceptedData) => ?RequestToApi;
export type Context = {
  rdfApi: string,
  rdfHeaders: Object,
  rdfInterceptor: ?Interceptor,
  rdfLoader: React$Node,
  rdfStore: Object,
  rdfTimeout: number,
}

export type ErrorContent = {
  request: XMLHttpRequest,
  response: String | Object,
}

export type Error = {
  content: ErrorContent,
  message: 'Something went wrong during the request',
  url?: string,
}

export type Method =
  'DELETE'
| 'FORM_DATA'
| 'GET'
| 'HEAD'
| 'PATCH'
| 'POST'
| 'PUT'
| 'TRACE'

export type DefaultProps = {
  body: Object,
  method: Method,
  params: Object,
}

export type Progress = {
  bubbles: boolean,
  cancelable: boolean,
  lengthComputable: boolean,
  loaded: number,
  target: EventTarget,
  total: number,
  type: string,
}

export type ReturnedData = {
  data?: Object,
  error?: Error,
  isOK?: boolean,
  request?: XMLHttpRequest,
  status?: number,
  store?: Object,
}

export type Props = {
  body?: Object,
  children?: React$StatelessFunctionalComponent<?ReturnedData | Error>,
  component?: React$ComponentType<?ReturnedData | Error>,
  headers?: Object,
  loader?: React$Node,
  method: Method,
  onError?: (?ReturnedData | Error) => void,
  onFetch?: (?ReturnedData | Error) => void,
  onLoad?: Function,
  onProgress?: (Progress) => void,
  onTimeout?: Function,
  onIntercept?: Interceptor,
  params?: Object,
  path?: string,
  refetchKey?: any,
  render?: React$StatelessFunctionalComponent<?ReturnedData | Error>,
  resultOnly?: boolean,
  timeout?: number,
  url?: string,
}

export type RequestToApi = {
  body?: Object,
  headers?: Object,
  method: Method,
  onProgress?: (Progress) => void,
  onTimeout?: Function,
  onIntercept?: ?Interceptor,
  params?: Object,
  url: string,
  timeout?: number,
}

export type InterceptedData = {
  currentParams: RequestToApi,
  request: XMLHttpRequest,
  status: number
}

export type Store = {
  subscribe: Function,
  dispatch: Function,
  getState: () => Object,
}

export type ProviderProps = {
  api: string,
  children: React$Node,
  headers?: Object,
  loader?: React$Node,
  onIntercept?: Interceptor,
  store?: Store,
  timeout?: number,
}

// PROPTYPES
export const methodShape = PropTypes.oneOf([
  'DELETE',
  'FORM_DATA',
  'GET',
  'HEAD',
  'PATCH',
  'POST',
  'PUT',
  'TRACE',
])

export const storeShape = PropTypes.shape({
  getState: PropTypes.func,
})
