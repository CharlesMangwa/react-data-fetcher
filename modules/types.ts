// import PropTypes from "prop-types";

import React from "react";

export type Interceptor = (interceptedData: IInterceptedData) => IRequestToApi | void;

export interface IContext {
  api?: string;
  headers?: {[s: string]: string};
  onIntercept?: Interceptor;
  loader?: React.ReactNode;
  store?: object;
  timeout?: number;
}

export interface IErrorContent {
  request: XMLHttpRequest;
  response: string | object;
}

export interface IError {
  content: IErrorContent;
  message: "Something went wrong during the request";
  url?: string;
}

export type Method =
  | "DELETE"
  | "FORM_DATA"
  | "GET"
  | "HEAD"
  | "PATCH"
  | "POST"
  | "PUT"
  | "TRACE";

export interface IReturnedData {
  data?: object;
  error?: IError;
  isOK?: boolean;
  request?: XMLHttpRequest;
  status?: number;
  store?: object;
}

export interface IDataFetcher {
  fetchData(): Promise<void>;
}

export interface IProps {
  body: {[s: string]: any};
  cancel?: boolean;
  children?: React.ReactNode;
  component?: React.ComponentType<IReturnedData | IError>;
  context: IContext;
  headers?: {[s: string]: string};
  loader?: React.ReactNode;
  method: Method;
  onError?: (data?: IReturnedData | IError) => void;
  onFetch?: (data?: IReturnedData | IError) => void;
  onIntercept?: Interceptor;
  onLoad?: () => void;
  onProgress?: (progressEvent: ProgressEvent) => void;
  onTimeout?: () => void;
  params?: {[s: string]: any};
  path?: string;
  refetchKey?: any;
  render?: React.SFC<IReturnedData | IError>;
  resultOnly?: boolean;
  timeout: number;
  url?: string;
}

export interface IRequestToApi {
  body?: {[s: string]: any};
  cancel?: boolean;
  headers?: {[s: string]: string};
  method: Method;
  onProgress?: (progressEvent: ProgressEvent) => void;
  onIntercept?: Interceptor;
  onTimeout?: () => void;
  params?: {[s: string]: any};
  url: string;
  timeout?: number;
}

export interface IInterceptedData {
  currentParams: IRequestToApi;
  request: XMLHttpRequest;
  status: number;
}

export interface IProviderProps {
  api: string;
  children: React.ReactNode;
  headers?: {[s: string]: string};
  loader?: React.ReactNode;
  onIntercept?: Interceptor;
  store?: object;
  timeout?: number;
}

/*
export const contextShape = PropTypes.shape({
  api: PropTypes.string,
  headers: PropTypes.object,
  loader: PropTypes.element,
  onIntercept: PropTypes.func,
  store: PropTypes.object,
  timeout: PropTypes.number,
});

export const methodShape = PropTypes.oneOf([
  "DELETE",
  "FORM_DATA",
  "GET",
  "HEAD",
  "PATCH",
  "POST",
  "PUT",
  "TRACE",
]);
*/
