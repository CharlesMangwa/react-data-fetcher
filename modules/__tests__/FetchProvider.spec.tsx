import React from "react";
import TestRenderer from "react-test-renderer";
import { createRenderer } from "react-test-renderer/shallow";

import { Fetch, FetchProvider } from "../index";
import { IError, IReturnedData} from "../types";

describe("A <FetchProvider />", () => {
  let fn;
  let renderer;

  beforeEach(() => {
    fn = jest.fn();
    renderer = createRenderer();
  });

  afterEach(() => jest.clearAllMocks());

  it("should render component children correctly", () => {
    expect.assertions(1);
    const component = TestRenderer.create(
      <FetchProvider
        value={{
          api: "https://api.github.com",
          headers: {
            "X-Nyan-Token": "superNyan",
          },
        }}
      >
        <Fetch path="/users/octocat">
          <div />
        </Fetch>
      </FetchProvider>,
    );
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("should propagate `store` correctly", () => {
    expect.assertions(1);
    const context = { api: "https://api.github.com", store: { cats: 42 } };
    const expectedData = {
      data: { cats: 42 },
      isOK: true,
    };
    const onFetch = (data?: IReturnedData | IError) => (receivedData = data || null);
    let receivedData;

    const component = TestRenderer.create(
      <FetchProvider value={context}>
        <Fetch path="store" onFetch={onFetch} />
      </FetchProvider>,
    );

    expect(receivedData).toMatchObject(expectedData);
  });
});
