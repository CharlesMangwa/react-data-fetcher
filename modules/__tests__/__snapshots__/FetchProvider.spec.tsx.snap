// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`A <FetchProvider /> should render component children correctly 1`] = `null`;
